src/uct/sm/scopy/base/scopy_iface.c

```diff
@@ -0,0 +1,92 @@
+{
+    uct_scopy_iface_t *iface = ucs_derived_of(tl_iface, uct_scopy_iface_t);
+
+    uct_base_iface_query(&iface->super.super, iface_attr);
+
+    /* default values for all shared memory transports */
+    iface_attr->cap.put.min_zcopy       = 0;
+    iface_attr->cap.put.max_zcopy       = SIZE_MAX;
+    iface_attr->cap.put.opt_zcopy_align = 1;
+    iface_attr->cap.put.align_mtu       = iface_attr->cap.put.opt_zcopy_align;
+    iface_attr->cap.put.max_iov         = iface->config.max_iov;
+
+    iface_attr->cap.get.min_zcopy       = 0;
+    iface_attr->cap.get.max_zcopy       = SIZE_MAX;
+    iface_attr->cap.get.opt_zcopy_align = 1;
+    iface_attr->cap.get.align_mtu       = iface_attr->cap.get.opt_zcopy_align;
+    iface_attr->cap.get.max_iov         = iface->config.max_iov;
+
+    iface_attr->device_addr_len         = uct_sm_iface_get_device_addr_len();
+    iface_attr->ep_addr_len             = 0;
+    iface_attr->max_conn_priv           = 0;
+        iface_attr->cap.flags               = UCT_IFACE_FLAG_GET_ZCOPY |
+                                                      UCT_IFACE_FLAG_PUT_ZCOPY |
+                                                      UCT_IFACE_FLAG_PENDING   |
+            UCT_IFACE_FLAG_CONNECT_TO_IFACE;
+        iface_attr->latency.overhead        = 80e-9; /* 80 ns */
+        iface_attr->latency.growth          = 0;
+}
+
+UCS_CLASS_INIT_FUNC(uct_scopy_iface_t, uct_scopy_iface_ops_t *ops, uct_md_h md,
+                    uct_worker_h worker, const uct_iface_params_t *params,
+                    const uct_iface_config_t *tl_config)
+{
+    uct_iface_ops_t iface_ops = {
+        .ep_put_zcopy             = ops->ep_put_zcopy,
+        .ep_get_zcopy             = ops->ep_get_zcopy,
+        .ep_pending_add           = (uct_ep_pending_add_func_t)ucs_empty_function_return_busy,
+        .ep_pending_purge         = (uct_ep_pending_purge_func_t)ucs_empty_function,
+        .ep_flush                 = uct_base_ep_flush,
+        .ep_fence                 = uct_sm_ep_fence,
+        .ep_create                = ops->ep_create,
+        .ep_destroy               = ops->ep_destroy,
+        .iface_flush              = uct_base_iface_flush,
+        .iface_fence              = uct_sm_iface_fence,
+        .iface_progress_enable    = ucs_empty_function,
+        .iface_progress_disable   = ucs_empty_function,
+        .iface_progress           = ucs_empty_function_return_zero,
+        .iface_close              = ops->iface_close,
+        .iface_query              = ops->iface_query,
+        .iface_get_address        = ops->iface_get_address,
+        .iface_get_device_address = uct_sm_iface_get_device_address,
+                .iface_is_reachable       = ops->iface_is_reachable
+    };
+    uct_scopy_iface_config_t *config = ucs_derived_of(tl_config,
+                                                      uct_scopy_iface_config_t);
+
+    UCS_CLASS_CALL_SUPER_INIT(uct_sm_iface_t, &iface_ops, md, worker, params, tl_config);
```

Comment:
you are passing a pointer to stack allocated ` iface_ops` obj. Base iface will not copy it, it just saves the pointer.
You'd need to declare static instance and use wrapper functions, which would call  `uct_scopy_iface_ops`  function stored in the scopy iface.

Reply:
no, I checked this.
BASE iface copied passed `ops` to the `ops` filed - https://github.com/openucx/ucx/blob/95c79a1a9e9363a66334c926129a8c2fd5e57050/src/uct/base/uct_iface.c#L436

Reply:
ok